### PR TITLE
[RFR] Update action chain runner - allow users to access action parameters inside "publish" scope

### PR DESCRIPTION
**Pull request by Kami**
Previously, user was unable to access action parameters inside the "publish" scope. This limitation made any things harder.

Here is an example of a work-flow where I use this functionality:

Metadata:

```yaml
 ...
  parameters:
    ip_address:
      type: "string"
      description: "IP address of a remote server to collect the informatiom from"
      required: true
```

Workflow:

```yaml
---
  chain:
    -
      name: "create_temporary_directory"
      ref: "core.remote_sudo"
      params:
        hosts: "{{ ip_address }}"
        cmd: "mktemp -d"
      publish:
        temp_dir: "{{ create_temporary_directory[parameters.ip_address].stdout }}"
      on-success: "retrieve_logged_users"

...
```

TODO:

- [x] Tests
- [x] Update changelog
- [x] Update docs

### CODE REVIEW

File: st2actions/st2actions/runners/actionchainrunner.py [hunk 1]
```diff
@@ -130,13 +130,16 @@ def pre_run(self):
...
-
         result = {'tasks': []}  # holds final result we store
         context_result = {}  # holds result which is used for the template context purposes
         top_level_error = None  # stores a reference to a top level error
         fail = True
         action_node = None
 
+        # Add action parameters to the context result so they can be accessed
+        # inside "publish"
+        context_result['parameters'] = action_parameters
```
Comment (Kami) [inline st2actions/st2actions/runners/actionchainrunner.py]: Since this is a "special" attribute maybe we should prefix it with `_` or similar. I'm open to suggestions.

Comment (manasdk) [inline st2actions/st2actions/runners/actionchainrunner.py]: similar to https://github.com/StackStorm/st2/blob/action_chain_runner_publish_access_to_action_params/st2actions/st2actions/runners/actionchainrunner.py#L39 ?

Comment (Kami) [inline st2actions/st2actions/runners/actionchainrunner.py]: Yep, I saw the results and I was wondering how and if we should prefix it.

I'm personally not a big fan of double underscore since `__` in Python means something internal which you shouldn't really touch. In the action parameters case, I want to prefix it to avoid potential name clashes and not because it's an "internal" / "secret" parameter.

Comment (Kami) [inline st2actions/st2actions/runners/actionchainrunner.py]: To clarify it - inside the `params" section, action parameters are accessible in a root scope (there is no prefix), so we could also drop the prefix all together in the `publish` section.

E.g.

```yaml
---
  chain:
    -
      name: "create_temporary_directory"
      ref: "core.remote_sudo"
      params:
        hosts: "{{ ip_address }}"  -> no prefix here
        cmd: "mktemp -d"
      publish:
        temp_dir: "{{ create_temporary_directory[parameters.ip_address].stdout }}" -> we probably shouldn't have a prefix here for consistency reasons
      on-success: "retrieve_logged_users"
```

**Comment by Kami** [comment]
As mentioned in https://github.com/StackStorm/st2/pull/1234#discussion_r26008026, I've decided to drop the attribute name prefix.

This way it's consistent with the existing code. Existing code already has a potential name clashes issue and solving it in just one case is worse than being consistent. We should consider this when designing an improved action chain runner and make sure we prefix the special attributes there.

**Comment by dzimine** [comment]
Nice!

**Comment by manasdk** [comment]
+1

**Comment by jfryman** [comment]
Oh man, and this just merged as I was writing comments...

Adding `parameters` is not consistent with how we operate in Mistral. The diverging behaviors could lead to confusion...

**Comment by Kami** [comment]
@jfryman Action parameters are already accessible everywhere else in the action chain (params section, vars section, etc.) except the `publish` section - this pr fixed that.

Not exactly sure how this affects the mistral runner though - probably I just don't have enough context on it.

**Comment by jfryman** [comment]
@Kami Certainly following the PR... let me try to explain again.

Everywhere else in ActionChain, I can address parameters without the prefix `parameters`.

But, you have this... 

> Existing code already has a potential name clashes issue and solving it in just one case is worse than being consistent

Is there no way around this in the short term? There is nothing with Mistral that is directly affected, but the behavior is just odd. I have to remember the `parameters` prefix only in a specific part of the ActionChain workflow (publish). That's ugly.

**Comment by Kami** [comment]
@jfryman I just had a look at the existing mistral workflow definition - I assume you are referencing how publish works there - you need to explicitly define which action parameter you want to be available using `input`?

If so, I see what you mean, but I don't think this is an issue - there is no such thing as input in the action chain runner right now so the behaviors can't be exactly the same.

**Comment by Kami** [comment]
@jfryman Sounds like we are on the same page - I dropped the "parameters" prefix in the last commit. It's probably just my wording which made it confusing :)

**Comment by jfryman** [comment]
@Kami AH! Great! Ok. Yeah, I was trying to parse the comments and code, and got confused. That was really my only concern. :heart:

**Comment by Kami** [comment]
Heh, np - and yeah, I was kinda talking and replying to myself so it was slightly confusing :)

**Comment by Kami** [comment]
Merged into 0.8.1 and master.
